src/client.js

```diff
@@ -0,0 +1,229 @@
+const Cabal = require('cabal-core')
+const CabalDetails = require('./cabal-details')
+const crypto = require('hypercore-crypto')
+const DatDns = require("dat-dns")
+const ram = require('random-access-memory')
+const memdb = require('memdb')
+const level = require('level')
+const path = require('path')
+const mkdirp = require("mkdirp")
+
+class Client {
+  constructor (opts) {
+    if (!(this instanceof Client)) return new Client(opts)
+    // This is redundant, but we might want to keep the cabal map around
+    // in the case the user has access to cabal instances
+    this._keyToCabal = {}
+    // maps a cabal-core instance to a CabalDetails object
+    this.cabals = new Map()
+    this.currentCabal = null
+    this.config = opts.config
+    this.maxFeeds = opts.maxFeeds || 1000
+
+    let cabalDnsOpts = {
+        hashRegex: /^[0-9a-f]{64}?$/i,
+        recordName: 'cabal',
+        protocolRegex: /^cabal:\/\/([0-9a-f]{64})/i,
+        txtRegex: /^"?cabalkey=([0-9a-f]{64})"?$/i,
+    }
+    // also takes opts.persistentCache which has a read and write function
+    //   read: async function ()   // aka cache lookup function
+    //   write: async function ()  // aka cache write function
+    if (opts.persistentCache) cabalDnsOpts.persistentCache = opts.persistentCache
+    this.cabalDns = DatDns(cabalDnsOpts)
+  }
+
+  static getDatabaseVersion () {
+    return Cabal.databaseVersion
+  }
+
+  static scrub (key) {
+    return key.replace('cabal://', '').replace('cbl://', '').replace('dat://', '').replace(/\//g, '')
+  }
+
+  resolveName (name, cb) {
+      return this.cabalDns.resolveName(name).then((key) => { 
+          if (!cb) return Client.scrub(key)
+          cb(Client.scrub(key)) 
+      })
+  }
+
+  createCabal () {
+    return this.addCabal(crypto.keyPair().publicKey.toString('hex'))
+  }
+
+  addCabal (key) {
+      let cabalPromise
+      if (typeof key === 'string') {
+          cabalPromise = this.resolveName(key).then((resolvedKey) => {
+            const {temp, dbdir} = this.config
+            const storage = temp ? ram : dbdir + resolvedKey
```

Comment:
Does this mean `opts.config.dbdir` needs to end in a forward slash? Does everything still work if no `dbdir` is given?